File: 01. Arrays/06. All Subarrays with sum k/main.py
# 🔹 Brute Force (O(N²))
def subarray_sum_brute(nums, k):
    count = 0
    for i in range(len(nums)):
        sum_ = 0
        for j in range(i, len(nums)):
            sum_ += nums[j]
            if sum_ == k:
                count += 1
    return count

# 🔹 Optimized (O(N)) - Prefix Sum + HashMap
def subarray_sum_optimized(nums, k):
    prefix_sum_count = {0: 1}
    count = prefix_sum = 0

    for num in nums:
        prefix_sum += num
        if prefix_sum - k in prefix_sum_count:
            count += prefix_sum_count[prefix_sum - k]
        prefix_sum_count[prefix_sum] = prefix_sum_count.get(prefix_sum, 0) + 1

    return count

# Main function
def main():
    nums = [1, 2, 3, 4, 2, -1, 2]
    k = 5

    print("Brute Force:", subarray_sum_brute(nums, k))
    print("Optimized:", subarray_sum_optimized(nums, k))

main()

File: 01. Arrays/06. All Subarrays with sum k/main.java
import java.util.*;

public class main {
    
    // 🔹 Brute Force (O(N²))
    public static int subarraySumBrute(int[] nums, int k) {
        int count = 0;
        for (int i = 0; i < nums.length; i++) {
            int sum = 0;
            for (int j = i; j < nums.length; j++) {
                sum += nums[j];
                if (sum == k) count++;
            }
        }
        return count;
    }

    // 🔹 Optimized (O(N)) - Prefix Sum + HashMap
    public static int subarraySumOptimized(int[] nums, int k) {
        Map<Integer, Integer> prefixSumCount = new HashMap<>();
        prefixSumCount.put(0, 1);

        int count = 0, prefixSum = 0;
        for (int num : nums) {
            prefixSum += num;
            if (prefixSumCount.containsKey(prefixSum - k)) {
                count += prefixSumCount.get(prefixSum - k);
            }
            prefixSumCount.put(prefixSum, prefixSumCount.getOrDefault(prefixSum, 0) + 1);
        }
        return count;
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 4, 2, -1, 2};
        int k = 5;

        System.out.println("Brute Force: " + subarraySumBrute(nums, k));
        System.out.println("Optimized: " + subarraySumOptimized(nums, k));
    }
}
